### PR TITLE
Serve static themes and javascript - Closes mumuki/mumuki-platform#224

### PR DESCRIPTION
**Pull request by rodri042**
Closes https://github.com/mumuki/mumuki-office/issues/16

### CODE REVIEW

**Comment by flbulgarelli** [comment]
@rodri042 could you please have a look at https://github.com/mumuki/mumuki-platform/issues/224 and fix this issue with this PR, too? I think we should follow the same strategy as with serving custom css - add the custom JS as organization field and then expose it as a resource that can be consumed from the other applications.

**Comment by flbulgarelli** [comment]
I'd name such field `extension_javascript` or something like that.

File: db/migrate/20170118010535_make_theme_not_nullable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class MakeThemeNotNullable < ActiveRecord::Migration[5.0]
+  def change
+    change_column :organizations, :theme_stylesheet, :string,  default: ''
```
Comment (flbulgarelli) [inline db/migrate/20170118010535_make_theme_not_nullable.rb]: Theme should be `text`

File: app/controllers/concerns/with_requested_organization.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  included do
+    unless Rails.application.config.consider_all_requests_local
+      rescue_from ActionController::RoutingError, with: :not_found
+    end
+    rescue_from ActiveRecord::RecordNotFound, with: :not_found
+    rescue_from Mumukit::Auth::UnauthorizedAccessError, with: :forbidden
+  end
+
+  def requested_organization
+    Organization.find_by_name(id_param) || raise(ActiveRecord::RecordNotFound)
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_requested_organization.rb]: `raise` does not need parenthesis

File: app/controllers/concerns/with_requested_organization.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+module WithRequestedOrganization
+  extend ActiveSupport::Concern
+
+  included do
+    unless Rails.application.config.consider_all_requests_local
+      rescue_from ActionController::RoutingError, with: :not_found
+    end
+    rescue_from ActiveRecord::RecordNotFound, with: :not_found
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_requested_organization.rb]: Why is this here?

File: app/controllers/concerns/with_requested_organization.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    rescue_from Mumukit::Auth::UnauthorizedAccessError, with: :forbidden
+  end
+
+  def requested_organization
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_requested_organization.rb]: Instead of providing a helper method here, I'd go for the rails way of providing a `set_organization!` and using it in controllers with a `before_action :set_organization, only: ...`

File: app/controllers/concerns/with_requested_organization.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+module WithRequestedOrganization
```
Comment (flbulgarelli) [inline app/controllers/concerns/with_requested_organization.rb]: :+1: to extracting repeated code with a mixin

File: app/controllers/themes_controller.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+class ThemesController < ApplicationController
+  include WithRequestedOrganization
```
Comment (flbulgarelli) [inline app/controllers/themes_controller.rb]: I'd simply name this mixin `WithOrganization`

Comment (flbulgarelli) [inline app/controllers/themes_controller.rb]: In laboratory we have similar mixins, and we have named that way.

File: app/controllers/themes_controller.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+
+  def show
+    organization = requested_organization
+    render inline: organization.theme_stylesheet, content_type: 'text/css'
```
Comment (flbulgarelli) [inline app/controllers/themes_controller.rb]: I'd like those assets to be served statically. Please document on this PR which will be the css path so that we can easily cache it through nginx. 

On the other hand, how could we deal with cache invalidation?

File: app/models/organization.rb [hunk 8]
```diff
@@ -14,7 +15,16 @@ def notify!
     Mumukit::Nuntius.notify_event!({organization: as_json}, 'OrganizationCreated')
   end
 
-  def default_values
+  private
+
+  def set_default_values
```
Comment (flbulgarelli) [inline app/models/organization.rb]: Methods with effect should always end with `!` - `set_default_values!`. The big exception is controller action names.

File: app/models/organization.rb [hunk 9]
```diff
@@ -14,7 +15,16 @@ def notify!
...
+    self.private ||= false
     self.logo_url ||= 'http://mumuki.io/logo-alt-large.png'
+    self.theme_stylesheet ||= ''
+  end
+
+  def compile_sass
```
Comment (flbulgarelli) [inline app/models/organization.rb]: `compile_sass!`

File: app/models/organization.rb [hunk 10]
```diff
@@ -14,7 +15,16 @@ def notify!
...
+    file = Tempfile.with(self.theme_stylesheet)
+    self.theme_stylesheet = Sass::Engine.for_file(file.path, syntax: :scss).render
```
Comment (flbulgarelli) [inline app/models/organization.rb]: Perhaps the SASS compilation concern could be extracted to a mixin

File: lib/core_ext/tempfile.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+class Tempfile
+  def self.with(content)
```
Comment (flbulgarelli) [inline lib/core_ext/tempfile.rb]: This method name is misleading. `with` methods typically take a block, open a resource, execute the block, and close the resource. However, this method is just writing. So I would call it `write! `

File: config/application.rb [hunk 12]
```diff
@@ -12,5 +12,7 @@ class Application < Rails::Application
     config.generators.javascripts = false
 
     config.generators.test_framework = :rspec
+
+    Dir[File.join(Rails.root, "lib", "core_ext", "*.rb")].each {|l| require l }
```
Comment (flbulgarelli) [inline config/application.rb]: Don't do this unless you have a really good reason. Put your extensions in the standard `initializers` directory.

File: lib/core_ext/tempfile.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+    file = Tempfile.new('tmp')
+    file.write content
+    file.close
```
Comment (flbulgarelli) [inline lib/core_ext/tempfile.rb]: :warning: you are not closing the file if the write method fails. You should use `ensure`

**Comment by flbulgarelli** [comment]
Looks good to me, but please have a look at my comments, particularly at the following code ruby/rails code conventions: 

* https://github.com/mumuki/mumuki-office/pull/18#discussion_r96872730
* https://github.com/mumuki/mumuki-office/pull/18#discussion_r96874640

File: app/controllers/concerns/with_requested_organization.rb (hunk 2 above)
Comment (rodri042) [inline app/controllers/concerns/with_requested_organization.rb]: selection_032
Tell that to RubyMine (?

**Comment by rodri042** [comment]
@flbulgarelli I think this is mergeable now. Take a look

File: config/initializers/ensure_assets_folder.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+def ensure_dir!(dir)
+    FileUtils.mkdir_p "#{Rails.public_path.to_s}/#{dir}"
+end
```
Comment (flbulgarelli) [inline config/initializers/ensure_assets_folder.rb]: Use `File.join`

File: app/models/concerns/with_static_assets.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def generate_javascript!
+    generate_asset! 'javascripts', 'extension_javascript'
+  end
+
+  def generate_asset!(directory, property)
+    url_property = "#{property}_url"
+    content = send property
+    path = "#{directory}/#{name}-#{SecureRandom.hex}"
```
Comment (flbulgarelli) [inline app/models/concerns/with_static_assets.rb]: This means that the hash will change even if you have no changes to the JS or SCSS

**Comment by flbulgarelli** [comment]
Looks good to me. Let's wait for @aguspina review

File: app/models/concerns/with_sass.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+  extend ActiveSupport::Concern
+
+  included do
+    before_save :compile_sass!
+  end
+
+  private
+
+  def compile_sass!
+    file = Tempfile.write!(self.theme_stylesheet)
```
Comment (aguspina) [inline app/models/concerns/with_sass.rb]: We should persists scss stylesheet, and write css in the static file @rodri042

Comment (rodri042) [inline app/models/concerns/with_sass.rb]: Done
